File: without-library/graph/components/cycle_detection.py
import sys

sys.setrecursionlimit(200000)

class CycleDetection:
    def __init__(self, n, adj):
        self.n = n
        self.adj = adj
        self.visited = [False] * (n + 1)
        # For directed
        self.rec_stack = [False] * (n + 1)
        
    def find_undirected(self, u, p=-1):
        self.visited[u] = True
        for v in self.adj[u]:
            if v == p: continue
            if self.visited[v]: return True
            if self.find_undirected(v, u): return True
        return False
        
    def find_directed(self, u):
        self.visited[u] = True
        self.rec_stack[u] = True
        for v in self.adj[u]:
            if not self.visited[v]:
                if self.find_directed(v): return True
            elif self.rec_stack[v]:
                return True
        self.rec_stack[u] = False
        return False

if __name__ == "__main__":
    pass

File: without-library/graph/components/cycle_detection.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> adj[200005];
bool visited[200005];
bool rec_stack[200005];

bool dfs_undirected(int u, int p) {
    visited[u] = true;
    for(size_t i=0; i<adj[u].size(); i++) {
        int v = adj[u][i];
        if(v == p) continue;
        if(visited[v]) return true;
        if(dfs_undirected(v, u)) return true;
    }
    return false;
}

bool dfs_directed(int u) {
    visited[u] = true;
    rec_stack[u] = true;
    for(size_t i=0; i<adj[u].size(); i++) {
        int v = adj[u][i];
        if(!visited[v]) {
            if(dfs_directed(v)) return true;
        } else if(rec_stack[v]) {
            return true;
        }
    }
    rec_stack[u] = false;
    return false;
}

int main() {
    return 0;
}
